airbyte/mcp/safe_mode.py

```diff
@@ -0,0 +1,171 @@
+# Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+"""Safe mode filtering for MCP tools.
+
+This module provides a decorator to tag tool functions with MCP annotations
+for deferred registration with safe mode filtering.
+"""
+
+from __future__ import annotations
+
+import os
+from collections.abc import Callable
+from typing import Any, Literal, TypeVar
+
+from airbyte.mcp._annotations import (
+    DESTRUCTIVE_HINT,
+    IDEMPOTENT_HINT,
+    OPEN_WORLD_HINT,
+    READ_ONLY_HINT,
+)
+
+
+F = TypeVar("F", bound=Callable[..., Any])
+
+AIRBYTE_CLOUD_MCP_READONLY_MODE = (
+    os.environ.get("AIRBYTE_CLOUD_MCP_READONLY_MODE", "").strip() == "1"
+)
+AIRBYTE_CLOUD_MCP_SAFE_MODE = os.environ.get("AIRBYTE_CLOUD_MCP_SAFE_MODE", "").strip() == "1"
+
+_REGISTERED_TOOLS: list[tuple[Callable[..., Any], dict[str, Any]]] = []
+
+
+class SafeModeError(Exception):
+    """Raised when a tool is blocked by safe mode restrictions."""
+
+    pass
+
+
+def should_register_cloud_tool(annotations: dict[str, Any]) -> bool:
```

Comment:
Rename to `should_register_cloud_tool`. You should be able to use domain as a condition before applying filters.